Build warning "differs in levels of indirection" in `npy_atomic.h` with MSVC

While reviewing a Windows [build log](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1051045&view=logs&j=2d03c380-16cc-53df-474d-81b5b5dc65ac&t=df791f6b-c996-56df-84d9-327e79a73f8d) from conda-forge (numpy 2.1.2, MSVC 19.29), I noticed a lot of warnings for `npy_atomic.h` like this:
```
%SRC_DIR%\numpy\_core\src\common\npy_atomic.h(56): warning C4047: 'return': 'void *' differs in levels of indirection from 'volatile uint64_t'
```

This is in the branch of `_MSC_VER`, which should be hit by MSVC and possibly (if `STDC_ATOMICS` isn't defined) by `clang-cl` and the Intel compilers:

https://github.com/numpy/numpy/blob/83c34d57e26e25f8ec69995bc2228f07ff21efab/numpy/_core/src/common/npy_atomic.h#L53-L56

--- a/numpy/_core/src/common/npy_atomic.h
+++ b/numpy/_core/src/common/npy_atomic.h
@@ -55,7 +55,7 @@ npy_atomic_load_ptr(const void *obj) {
 #if defined(_M_X64) || defined(_M_IX86)
     return (void *)*(volatile uint64_t *)obj;
 #elif defined(_M_ARM64)
-    return (void *)(uintptr_t)__ldar64((unsigned __int64 volatile *)obj);
+    return (void *)__ldar64((unsigned __int64 volatile *)obj);
 #endif
 #else
 #if defined(_M_X64) || defined(_M_IX86)

--- a/numpy/_core/src/common/npy_atomic.h
+++ b/numpy/_core/src/common/npy_atomic.h
@@ -59,7 +59,7 @@ npy_atomic_load_ptr(const void *obj) {
 #endif
 #else
 #if defined(_M_X64) || defined(_M_IX86)
-    return (void *)(uintptr_t)*(volatile uint32_t *)obj;
+    return (void *)(volatile uint32_t *)obj;
 #elif defined(_M_ARM64)
     return (void *)(uintptr_t)__ldar32((unsigned __int32 volatile *)obj);
 #endif

--- a/numpy/_core/src/common/npy_atomic.h
+++ b/numpy/_core/src/common/npy_atomic.h
@@ -37,7 +37,7 @@ npy_atomic_load_uint8(const npy_uint8 *obj) {
     return (npy_uint8)atomic_load((const _Atomic(uint8_t)*)obj);
 #elif defined(MSC_ATOMICS)
 #if defined(_M_X64) || defined(_M_IX86)
-    return (void *)(uintptr_t)*(volatile uint64_t *)obj;
+    return *(volatile npy_uint8 *)obj;
 #else // defined(_M_ARM64)
     return (npy_uint8)__ldar8((unsigned __int8 volatile *)obj);
 #endif
@@ -53,7 +53,7 @@ npy_atomic_load_ptr(const void *obj) {
 #elif defined(MSC_ATOMICS)
 #if SIZEOF_VOID_P == 8
 #if defined(_M_X64) || defined(_M_IX86)
-    return *(volatile uint64_t *)obj;
+    return (void *)(uintptr_t)*(volatile uint64_t *)obj;
 #elif defined(_M_ARM64)
     return (uint64_t)__ldar64((unsigned __int64 volatile *)obj);
 #endif

--- a/numpy/_core/src/common/npy_atomic.h
+++ b/numpy/_core/src/common/npy_atomic.h
@@ -37,7 +37,7 @@ npy_atomic_load_uint8(const npy_uint8 *obj) {
     return (npy_uint8)atomic_load((const _Atomic(uint8_t)*)obj);
 #elif defined(MSC_ATOMICS)
 #if defined(_M_X64) || defined(_M_IX86)
-    return *(volatile npy_uint8 *)obj;
+    return (void *)(uintptr_t)*(volatile uint64_t *)obj;
 #else // defined(_M_ARM64)
     return (npy_uint8)__ldar8((unsigned __int8 volatile *)obj);
 #endif

--- a/numpy/_core/src/common/npy_atomic.h
+++ b/numpy/_core/src/common/npy_atomic.h
@@ -53,7 +53,7 @@ npy_atomic_load_ptr(const void *obj) {
 #elif defined(MSC_ATOMICS)
 #if SIZEOF_VOID_P == 8
 #if defined(_M_X64) || defined(_M_IX86)
-    return (void *)(uintptr_t)*(volatile uint64_t *)obj;
+    return (void *)*(volatile uint64_t *)obj;
 #elif defined(_M_ARM64)
     return (void *)(uintptr_t)__ldar64((unsigned __int64 volatile *)obj);
 #endif

--- a/numpy/_core/src/common/npy_atomic.h
+++ b/numpy/_core/src/common/npy_atomic.h
@@ -55,13 +55,13 @@ npy_atomic_load_ptr(const void *obj) {
 #if defined(_M_X64) || defined(_M_IX86)
     return (void *)(uintptr_t)*(volatile uint64_t *)obj;
 #elif defined(_M_ARM64)
-    return (uint64_t)__ldar64((unsigned __int64 volatile *)obj);
+    return (void *)(uintptr_t)__ldar64((unsigned __int64 volatile *)obj);
 #endif
 #else
 #if defined(_M_X64) || defined(_M_IX86)
-    return *(volatile uint32_t *)obj;
+    return (void *)(uintptr_t)*(volatile uint32_t *)obj;
 #elif defined(_M_ARM64)
-    return (uint32_t)__ldar32((unsigned __int32 volatile *)obj);
+    return (void *)(uintptr_t)__ldar32((unsigned __int32 volatile *)obj);
 #endif
 #endif
 #elif defined(GCC_ATOMICS)

--- a/numpy/_core/src/common/npy_atomic.h
+++ b/numpy/_core/src/common/npy_atomic.h
@@ -59,7 +59,7 @@ npy_atomic_load_ptr(const void *obj) {
 #endif
 #else
 #if defined(_M_X64) || defined(_M_IX86)
-    return (void *)(volatile uint32_t *)obj;
+    return (void *)*(volatile uint32_t *)obj;
 #elif defined(_M_ARM64)
     return (void *)__ldar32((unsigned __int32 volatile *)obj);
 #endif

--- a/numpy/_core/src/common/npy_atomic.h
+++ b/numpy/_core/src/common/npy_atomic.h
@@ -61,7 +61,7 @@ npy_atomic_load_ptr(const void *obj) {
 #if defined(_M_X64) || defined(_M_IX86)
     return (void *)(volatile uint32_t *)obj;
 #elif defined(_M_ARM64)
-    return (void *)(uintptr_t)__ldar32((unsigned __int32 volatile *)obj);
+    return (void *)__ldar32((unsigned __int32 volatile *)obj);
 #endif
 #endif
 #elif defined(GCC_ATOMICS)

